<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.1.179">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="SW">

<title>Genomics Workshop 2022 - Day 2</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Genomics Workshop 2022</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html">Home</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./Day1.html">Day 1</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./Day2.html" aria-current="page">Day 2</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./Day3.html">Day 3</a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#post-assembly-qc" id="toc-post-assembly-qc" class="nav-link" data-scroll-target="#post-assembly-qc">5. Post-Assembly QC</a>
  <ul class="collapse">
  <li><a href="#assembly-graphs-and-genome-size." id="toc-assembly-graphs-and-genome-size." class="nav-link" data-scroll-target="#assembly-graphs-and-genome-size.">5.1. Assembly graphs and genome size.</a></li>
  <li><a href="#k-mer-content" id="toc-k-mer-content" class="nav-link" data-scroll-target="#k-mer-content">5.2. k-mer content</a></li>
  <li><a href="#gene-content" id="toc-gene-content" class="nav-link" data-scroll-target="#gene-content">5.2. Gene content</a></li>
  <li><a href="#quast-report" id="toc-quast-report" class="nav-link" data-scroll-target="#quast-report">5.4. Quast report</a></li>
  </ul></li>
  <li><a href="#more-complex-genomes" id="toc-more-complex-genomes" class="nav-link" data-scroll-target="#more-complex-genomes">6. More complex genomes</a></li>
  <li><a href="#genome-alignment" id="toc-genome-alignment" class="nav-link" data-scroll-target="#genome-alignment">7. Genome alignment</a></li>
  <li><a href="#rna-seq-gene-annotation" id="toc-rna-seq-gene-annotation" class="nav-link" data-scroll-target="#rna-seq-gene-annotation">8. RNA-seq gene annotation</a>
  <ul class="collapse">
  <li><a href="#data-download" id="toc-data-download" class="nav-link" data-scroll-target="#data-download">8.1. Data download</a></li>
  <li><a href="#mapping" id="toc-mapping" class="nav-link" data-scroll-target="#mapping">8.2. Mapping</a></li>
  <li><a href="#transcript-assembly" id="toc-transcript-assembly" class="nav-link" data-scroll-target="#transcript-assembly">8.3. Transcript assembly</a></li>
  </ul></li>
  <li><a href="#homology-gene-annotation" id="toc-homology-gene-annotation" class="nav-link" data-scroll-target="#homology-gene-annotation">9. Homology gene annotation</a></li>
  <li><a href="#homology-gene-clustering" id="toc-homology-gene-clustering" class="nav-link" data-scroll-target="#homology-gene-clustering">10. Homology gene clustering</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Day 2</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>SW </p>
          </div>
  </div>
    
    
  </div>
  

</header>

<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>In the previous session, we used some different data types produced by different technologies to reconstruct the full DNA complement of a target organism. Today, we will do some more work on the resulting genomes and try and assess how good our constructions are. We will use some concepts covered in the last session, such as k-mers, but we will also start to incorporate information about functional elements such as genes. Before we do that, it’s important to recap on what a genome contains, and how it functions.</p>
<p>The DNA sequence of an organism is sometimes referred to as the data storage structure of living organisms. This storage contains a dizzying variety of features: Their are sequences that can be transcribed into various species of RNA. Some of these RNA molecules can be active in their own right, such as micro-RNA or CRISPR sequences. Other transcribed sequences go on to be translated into strings of amino-acids, that ultimately produce functional proteins.</p>
<section id="figure-1.-the-central-dogma-of-molecular-biology" class="level5">
<h5 class="anchored" data-anchor-id="figure-1.-the-central-dogma-of-molecular-biology">Figure 1. The central dogma of molecular biology</h5>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Screenshot%202022-10-27%20at%2015.26.47.png" class="img-fluid figure-img"></p>
</figure>
</div>
<p>Many regions of DNA are not typically converted into other types of biomolecule, however they can still confer important structural properties (e.g.&nbsp;centromeres or telomeres), or provide regulation of the regions that are transcribed (e.g.&nbsp;promoters or enhancers). Genomes are also full of viral and non-viral mobile DNA sequences such as transposable elements, the differing activities of these mobile sequences are fascinating in their own right, and can have mutagenic activities, with important implications for evolutionary biology.</p>
<p>Genome annotation refers broadly to the identification of the various features in a genome sequence. Additional data is often used to achieve this. Rather than generate sequencing read data from genomic DNA, we can use the same technology to capture a sample of an organisms RNA. “Mapping” this data (aligning the RNA sequencing reads to a genome assembly), can provide evidence for the locations of transcribed sequence including their, start, end and gene splicing events. Based on the sequences within these annotations we can then predict resulting amino-acid sequences, and in some cases the three-dimensional structure of the resulting proteins.</p>
<p>Here is a genome browser displaying annotations for an E. coli reference genome, have a go at zooming in to see the various annotations on the genome, and then further to the nucleotide sequence:</p>
<iframe id="iframe_NCBI" width="708" height="354" src="https://www.ncbi.nlm.nih.gov/projects/sviewer/embedded_iframe.html?iframe=iframe_NCBI&amp;id=NC_000913.3&amp;tracks=[key:sequence_track,name:Sequence,display_name:Sequence,id:STD649220238,annots:Sequence,ShowLabel:false,ColorGaps:false,shown:true,order:1][key:gene_model_track,name:Genes,display_name:Genes,id:STD3194982005,annots:Unnamed,Options:MergeAll,CDSProductFeats:false,NtRuler:true,AaRuler:true,HighlightMode:2,ShowLabel:true,shown:true,order:5]&amp;assm_context=GCF_000005845.2&amp;v=1:4641652&amp;c=800080&amp;select=null&amp;slim=0&amp;appname=no_appname" onload="if(!window._SViFrame){_SViFrame=true;window.addEventListener('message',function(e){if(e.origin=='https://www.ncbi.nlm.nih.gov' &amp;&amp; !isNaN(e.data.h))document.getElementById(e.data.f).height=parseInt(e.data.h);});}">
</iframe>
</section>
</section>
<section id="post-assembly-qc" class="level2" data-link="5. Post-Assembly QC">
<h2 data-link="5. Post-Assembly QC" class="anchored" data-anchor-id="post-assembly-qc">5. Post-Assembly QC</h2>
<p>Now that we have some assembled genomes from the previous session, we will explore them in some more detail. This should help you understand the strengths and weaknesses of the different data-types, and also how to assess different aspects of genome assembly quality.</p>
<p>Both Canu and Spades should have produced directories full of various files. Hifiasm should also produce output files, but for this program we need to do an additional step to generate a “fasta” file from the “gfa” file.</p>
<pre><code>awk '/^S/{print "&gt;"$2;print $3}' hifi_ecoli.p_ctg.gfa &gt; hifi_ecoli.p_ctg.fa</code></pre>
<section id="assembly-graphs-and-genome-size." class="level3">
<h3 class="anchored" data-anchor-id="assembly-graphs-and-genome-size.">5.1. Assembly graphs and genome size.</h3>
<p>We can visualise the assembly graphs for these three different datasets using the software Bandage. We use a specific file-type produced by some (but not all) genome assembly programs, known as a Graphical Fragment Assembly (.gfa) file. This format documents stretches of assembled sequence and also the relationship between contiguous runs, that may have multiple valid paths due to the presence of repetitive sequence or heterozygosity.</p>
<p>To use bandage in Window:</p>
<ol type="1">
<li><p>Use “scp” to copy the gfa or fastg files from the three assemblies to your local machine.</p></li>
<li><p>Download the program from this link https://github.com/rrwick/Bandage/releases/download/v0.8.1/Bandage_Windows_v0_8_1.zip</p></li>
<li><p>Open the zip archive and double click on the Bandage.exe file.</p></li>
<li><p>A window should open.</p></li>
<li><p>You can click File, then load graph, and you should be able to select a gfa file from one of your assemblies.</p></li>
<li><p>Click “Draw graph”</p></li>
</ol>
<p>You can load bandage in Linux by:</p>
<ol type="1">
<li><p>typing the command “Bandage” into the unix command prompt.</p></li>
<li><p>A window should open.</p></li>
<li><p>You can click File, then load graph, and you should be able to select a gfa file from one of your assemblies.</p></li>
</ol>
<section id="figure-3-a-visualisation-of-the-gfa-file-produced-by-spades-using-the-short-illumina-read-data.-we-can-see-that-there-is-a-tangle-of-valid-paths-and-also-that-the-total-length-is-higher-than-the-size-estimated-from-our-earlier-k-mer-analysis.-there-are-long-stretches-of-contiguous-sequence-which-is-useful-genomic-data-but-the-unresolved-regions-of-the-graph-may-inhibit-some-types-of-analysis.-we-would-refer-to-this-as-a-draft-genome." class="level5">
<h5 class="anchored" data-anchor-id="figure-3-a-visualisation-of-the-gfa-file-produced-by-spades-using-the-short-illumina-read-data.-we-can-see-that-there-is-a-tangle-of-valid-paths-and-also-that-the-total-length-is-higher-than-the-size-estimated-from-our-earlier-k-mer-analysis.-there-are-long-stretches-of-contiguous-sequence-which-is-useful-genomic-data-but-the-unresolved-regions-of-the-graph-may-inhibit-some-types-of-analysis.-we-would-refer-to-this-as-a-draft-genome.">Figure 3: A visualisation of the gfa file produced by Spades using the short Illumina read data. We can see that there is a tangle of valid paths, and also that the total length is higher than the size estimated from our earlier k-mer analysis. There are long stretches of contiguous sequence, which is useful genomic data, but the unresolved regions of the graph may inhibit some types of analysis. We would refer to this as a draft genome.</h5>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Screenshot%202022-10-13%20at%2014.30.49.png" class="img-fluid figure-img"></p>
</figure>
</div>
</section>
<section id="figure-4-in-contrast-here-is-a-visualisation-of-the-unitig-gfa-file-produced-by-canu-using-the-nanopore-data.-we-can-see-that-for-the-most-part-there-is-a-single-clean-path-node-and-that-the-total-length-is-roughly-the-size-we-expect.-after-checking-the-final-contig-sequence-file-fasta-the-loop-structure-is-resolved-into-a-single-circular-genome." class="level5">
<h5 class="anchored" data-anchor-id="figure-4-in-contrast-here-is-a-visualisation-of-the-unitig-gfa-file-produced-by-canu-using-the-nanopore-data.-we-can-see-that-for-the-most-part-there-is-a-single-clean-path-node-and-that-the-total-length-is-roughly-the-size-we-expect.-after-checking-the-final-contig-sequence-file-fasta-the-loop-structure-is-resolved-into-a-single-circular-genome.">Figure 4: In contrast, here is a visualisation of the “unitig” gfa file produced by canu using the nanopore data. We can see that, for the most part there is a single clean path (node), and that the total length is roughly the size we expect. After checking the final contig sequence file (fasta), the loop structure is resolved into a single circular genome.</h5>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Screenshot%202022-10-14%20at%2014.02.47.png" class="img-fluid figure-img"></p>
</figure>
</div>
</section>
<section id="figure-5-here-is-a-visualisation-of-the-gfa-file-produced-by-hifiasm-using-the-pacbio-hifi-dataset.-we-can-see-that-this-data-is-able-to-resolve-the-genome-into-a-single-closed-loop." class="level5">
<h5 class="anchored" data-anchor-id="figure-5-here-is-a-visualisation-of-the-gfa-file-produced-by-hifiasm-using-the-pacbio-hifi-dataset.-we-can-see-that-this-data-is-able-to-resolve-the-genome-into-a-single-closed-loop.">Figure 5: Here is a visualisation of the gfa file produced by hifiasm using the PacBio HiFi dataset. We can see that this data is able to resolve the genome into a single closed loop.</h5>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Screenshot%202022-10-13%20at%2014.29.47.png" class="img-fluid figure-img"></p>
</figure>
</div>
</section>
</section>
<section id="k-mer-content" class="level3">
<h3 class="anchored" data-anchor-id="k-mer-content">5.2. k-mer content</h3>
<p>Another way to check the quality of a genome assembly, without having to rely on genes, is to compare the k-mer content that we saw in the raw reads to the k-mers present in the finished genome assembly. We can use the k-mer analysis toolkit to do this.</p>
<pre><code>kat comp -t 16 -o SRR18106304_vs_spades SRR18106304.fastq ecoli_hiseq_spades/scaffolds.fasta</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Screenshot%202022-10-24%20at%2016.42.40.png" class="img-fluid figure-img"></p>
</figure>
</div>
<p>Here we can see the same distribution we observed earlier, however now we have colours added, which indicate how many times k-mers occur in the assembled genome. We can see here on the left hand side that many of the error k-mers are shown in black. This means the k-mers are not present in the assembly, which is good news. However, we can also see that some of these error k-mers are shown in red, which means there are some sequence errors still present in the draft assembly.</p>
<p>In comparison here is the same plot for the HiFi assembly.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Screenshot%202022-10-24%20at%2016.40.24.png" class="img-fluid figure-img"></p>
</figure>
</div>
<p>Now we can see that there are much less k-mers in the left hand peak present in the assembly. although we can see that a small number of k-mers in the right hand peak contains a small amount of purple duplicated k-mers!</p>
</section>
<section id="gene-content" class="level3">
<h3 class="anchored" data-anchor-id="gene-content">5.2. Gene content</h3>
<p>We can check the quality of an assembly, by checking whether “conserved” genes are present in our assemblies (see paper for details). Using the software BUSCO (Benchmarking Universal Single-Copy Orthologs), we can generate some information about whether genes are present or absent, fragmented across contigs, or whether they’re present in the right number.</p>
<p>Here we are performing a specific type of genome annotation, which attempts to identify a specific set of conserved genes, that are often only present in a single copy. If some of these genes are missing, it provides some evidence that our genome assembly may not be complete. Alternatively if these genes are found more than once, it provides some evidence that our assembly may have incorrectly duplicated some genomic regions.</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
BUSCO is currently not working, an example output will be available in /home/taller-2019/Workshop_materials/
</div>
</div>
<div class="callout-body-container callout-body">

</div>
</div>
<pre><code>##Bash/command line code 
busco -m genome -i /scratch/ecoli_hiseq_spades/scaffolds.fasta -o Hiseq_busco --auto-lineage-prok</code></pre>
<p>Have a browse through the output files in the output folder Hiseq_busco and see if you can find out how complete this genome assembly is. You can also generate an analysis for the other genomes you assembled yesterday (make sure to change the name of the output file!) and compare the results.</p>
</section>
<section id="quast-report" class="level3">
<h3 class="anchored">5.4. Quast report</h3>
<p>Before we move on to some other checks we are going to run the program quast, which performs a range of checks for you. We can run this as indicated below. For the purposes of demonstration we will use a reference genome to check our data against, but for a novel genome assembly we would not have this data.</p>
<p>We are going to run this software with another form of annotation, to help us assess the quality of our genome assemblies. The Glimmer method is specifically designed to detect prokaryotic genes, without the need for additional data.</p>
<ol type="1">
<li><p>First find an E. coli k-12 reference genome on NCBI (https://www.ncbi.nlm.nih.gov/), as shown below.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Screenshot%202022-10-26%20at%2010.26.34.png" class="img-fluid figure-img"></p>
</figure>
</div></li>
<li><p>Select one of the resulting genome assemblies.</p></li>
<li><p>Now click on the link on the right hand side of the page labelled “FTP directory for RefSeq assembly”</p></li>
<li><p>Here you can see a range of information including the genomic sequence files annotations and other associated statistics.</p></li>
<li><p>Right click on the file that looks like this “GCF_905147045.1_ilAglIoxx1.1_genomic.fna.gz” and copy the link.</p></li>
<li><p>Now in your Unix terminal navigate to your home directory and paste the link after a “wget command”, in the same format as below (make sure to paste in the correct link that you’ve copied).</p>
<p>wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_example/GCF_example_genomic.fna.gz</p></li>
<li><p>Now you are ready to run quast with your three previously assembled genomes.</p></li>
</ol>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Load both modules for quast</p>
<section id="module-load-anaconda24.2.0" class="level2">
<h2 class="anchored" data-anchor-id="module-load-anaconda24.2.0">module load anaconda2/4.2.0</h2>
</section>
<section id="module-load-quast4.6.0" class="level2">
<h2 class="anchored" data-anchor-id="module-load-quast4.6.0">module load quast/4.6.0</h2>
</section>
</div>
</div>
<pre><code>quast.py -R GCF_000005845.2_ASM584v2_genomic.fna --glimmer --output-dir test_quast ./Hifi_ecoli.fa ./ecoli-oxford/ecoli.contigs.fasta ./ecoli_hiseq_spades/scaffolds.fasta  </code></pre>
<ol start="8" type="1">
<li><p>Have a look at the quast report by loading the html file.</p>
<p>firefox test_quast/report.html</p></li>
</ol>
</section>
</section>
<section id="more-complex-genomes" class="level2" data-link="6. More complex genomes">
<h2 data-link="6. More complex genomes" class="anchored" data-anchor-id="more-complex-genomes">6. More complex genomes</h2>
<p>So far we have only considered a haploid bacterial genome, with a very small overall size. In such simple genomes an assembly graphs may suggest multiple paths through a given sequence due to repeating sequence, but we know that there is likely only one biologically real path through the graph. In diploid (and polyploid) organisms, this becomes more complicated because there are more than one real biological paths through the graph. These alternate paths are due to the presence of sister chromatids, that are not identical to each other, due to the presence of genetic variation, or heterozygosity. A key factor in assembling high quality genome assemblies, is the “phasing” of the chromosomes. It can be difficult to distinguish whether a given heterozygous region belongs to one chromosome copy or the other. Variation found on the same chromosome copy is in “linkage” or can be said to have the same “phase”. A perfect genome assembly would reconstruct the variation from each chromosome perfectly as found in the original organism, however in practice this can be difficult to achieve. The figures below should help to explain this problem and explain potential solutions.</p>
<section id="figure-6-a-common-result-from-sequencing-diploid-genomes-is-the-presence-of-redundant-haplotypes-in-the-final-assembly-particularly-for-regions-with-have-a-high-sequence-divergence.-here-we-can-see-homozygous-regions-in-orange-and-a-reconstructed-reference-contig-which-contains-a-mosaic-of-haplotypes-from-two-different-sister-chromatids-indicated-in-yellow-and-blue." class="level5">
<h5 class="anchored" data-anchor-id="figure-6-a-common-result-from-sequencing-diploid-genomes-is-the-presence-of-redundant-haplotypes-in-the-final-assembly-particularly-for-regions-with-have-a-high-sequence-divergence.-here-we-can-see-homozygous-regions-in-orange-and-a-reconstructed-reference-contig-which-contains-a-mosaic-of-haplotypes-from-two-different-sister-chromatids-indicated-in-yellow-and-blue.">Figure 6: A common result from sequencing diploid genomes is the presence of redundant haplotypes in the final assembly, particularly for regions with have a high sequence divergence. Here we can see homozygous regions in orange, and a reconstructed reference contig which contains a mosaic of haplotypes from two different sister chromatids (indicated in yellow and blue).</h5>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Screenshot%202022-10-24%20at%2017.25.47.png" class="img-fluid figure-img"></p>
</figure>
</div>
</section>
<section id="figure-7-one-method-of-resolving-heterozygosity-in-a-genome-sequencing-project-is-to-gather-parental-data-and-use-this-to-partition-the-chromatids-inherited-by-the-f1-offspring.-in-this-example-we-generated-long-reads-for-an-f1-individual-and-short-illumina-reads-for-the-parents.-we-were-able-to-use-the-illumina-reads-to-separate-the-f1-pacbio-reads-into-two-groups-for-assembly." class="level5">
<h5 class="anchored" data-anchor-id="figure-7-one-method-of-resolving-heterozygosity-in-a-genome-sequencing-project-is-to-gather-parental-data-and-use-this-to-partition-the-chromatids-inherited-by-the-f1-offspring.-in-this-example-we-generated-long-reads-for-an-f1-individual-and-short-illumina-reads-for-the-parents.-we-were-able-to-use-the-illumina-reads-to-separate-the-f1-pacbio-reads-into-two-groups-for-assembly.">Figure 7: One method of resolving heterozygosity in a genome sequencing project is to gather parental data and use this to partition the chromatids inherited by the F1 offspring. In this example, we generated long reads for an F1 individual and short Illumina reads for the parents. We were able to use the Illumina reads to separate the F1 PacBio reads into two groups for assembly.</h5>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Screenshot%202022-10-24%20at%2017.37.37.png" class="img-fluid figure-img"></p>
</figure>
</div>
</section>
<section id="figure-7-another-way-to-utilise-parental-data-is-to-use-the-k-mers-during-the-construction-of-the-assembly-graph-itself.-specifically-we-can-identify-bubble-structures-that-are-likely-to-represent-heterozygous-regions-and-look-for-parent-specific-k-mers-in-the-alternate-paths.-crucially-this-approach-requires-high-accuracy-long-reads-such-as-hifi-data-to-work-successfully." class="level5">
<h5 class="anchored" data-anchor-id="figure-7-another-way-to-utilise-parental-data-is-to-use-the-k-mers-during-the-construction-of-the-assembly-graph-itself.-specifically-we-can-identify-bubble-structures-that-are-likely-to-represent-heterozygous-regions-and-look-for-parent-specific-k-mers-in-the-alternate-paths.-crucially-this-approach-requires-high-accuracy-long-reads-such-as-hifi-data-to-work-successfully.">Figure 7: Another way to utilise parental data is to use the k-mers during the construction of the assembly graph itself. Specifically, we can identify bubble structures that are likely to represent heterozygous regions and look for parent specific k-mers in the alternate paths. Crucially, this approach requires high accuracy long reads such as HiFi data to work successfully.</h5>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Screenshot%202022-10-24%20at%2017.37.51.png" class="img-fluid figure-img"></p>
</figure>
</div>
<section id="section" class="level6">
<h6 class="anchored" data-anchor-id="section"></h6>
</section>
</section>
</section>
<section id="genome-alignment" class="level2">
<h2 class="anchored" data-anchor-id="genome-alignment">7. Genome alignment</h2>
<p>We are going to download some datasets, you may choose two species from the following list to align:</p>
<table class="table">
<thead>
<tr class="header">
<th>Name (common)</th>
<th>NCBI RefSeq accession</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Melitaea cinxia (Glanville fritillary)</td>
<td><a href="https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/905/220/565/GCF_905220565.1_ilMelCinx1.1/">GCF_905220565.1</a></td>
</tr>
<tr class="even">
<td>Pararge aegeria (specked wood butterfly)</td>
<td><a href="https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/905/163/445/GCF_905163445.1_ilParAegt1.1/">GCF_905163445.1</a></td>
</tr>
<tr class="odd">
<td>Nymphalis io (European peacock)</td>
<td><a href="https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/905/147/045/GCF_905147045.1_ilAglIoxx1.1/">GCF_905147045.1</a></td>
</tr>
<tr class="even">
<td>Vanessa cardui (painted lady)</td>
<td><a href="https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/905/220/365/GCF_905220365.1_ilVanCard2.1/">GCF_905220365.1</a></td>
</tr>
<tr class="odd">
<td>Leptidea sinapis (wood white)</td>
<td><a href="https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/905/404/315/GCF_905404315.1_ilLepSina1.1/">GCF_905404315.1</a></td>
</tr>
<tr class="even">
<td>Maniola jurtina (meadow brown)</td>
<td><a href="https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/905/333/055/GCF_905333055.1_ilManJurt1.1/">GCF_905333055.1</a></td>
</tr>
<tr class="odd">
<td>Vanessa atalanta (red admiral)</td>
<td><a href="https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/905/147/765/GCF_905147765.1_ilVanAtal1.2/">GCF_905147765.1</a></td>
</tr>
<tr class="even">
<td>Plutella xylostella (diamondback moth)</td>
<td><a href="https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/932/276/165/GCF_932276165.1_ilPluXylo3.1/">GCF_932276165.1</a></td>
</tr>
<tr class="odd">
<td>Pieris rapae (cabbage white)</td>
<td><a href="https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/905/147/795/GCF_905147795.1_ilPieRapa1.1/">GCF_905147795.1</a></td>
</tr>
<tr class="even">
<td>Pieris brassicae (large cabbage white)</td>
<td><a href="https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/905/147/105/GCF_905147105.1_ilPieBrab1.1/">GCF_905147105.1</a></td>
</tr>
<tr class="odd">
<td>Colias croceus (clouded yellow)</td>
<td><a href="https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/905/220/415/GCF_905220415.1_ilColCroc2.1/">GCF_905220415.1</a></td>
</tr>
<tr class="even">
<td>Papilio machaon (common yellow swallowtail)</td>
<td><a href="https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/912/999/745/GCF_912999745.1_ilPapMach1.1/">GCF_912999745.1</a></td>
</tr>
<tr class="odd">
<td>Pieris napi (green-veined white)</td>
<td><a href="https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/905/475/465/GCF_905475465.1_ilPieNapi1.2/">GCF_905475465.1</a></td>
</tr>
<tr class="even">
<td>Aricia agestis (brown argus)</td>
<td><a href="https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/905/147/365/GCF_905147365.1_ilAriAges1.1/">GCF_905147365.1</a></td>
</tr>
</tbody>
</table>
<p>In section 5.4 we downloaded an E. coli reference genome. You will need to perform this same process for each of your chosen species.</p>
<p>Once you have the genomic fasta files downloaded, we can try out a pairwise genomic alignment between two of your species. We’ll use the tool minimap2, which was designed to map long sequences to a genome, such as pacbio reads, however using specific parameters we can also use it to perform alignments between species that are not too distantly related (i.e.&nbsp;within the same Family, or Order).</p>
<p>Using two fasta files we can compute a “paf” alignment file (here I use Aricia agestis and Vanessa cardui), like so:</p>
<pre><code>~/bin/minimap2/minimap2 -PD -x asm5 -t 16 GCF_905147365.1_ilAriAges1.1_genomic.fna GCF_905220365.1_ilVanCard2.1_genomic.fna  &gt; AriAges_VanCard.paf</code></pre>
<p>We can then use an R script to convert this alignment into a dotplot style figure. This type of figure represents a query genome on one axis and a target genome on the other axis. Dots correspond to matches between the two sets of sequences. Some dots will appear across many different chromosomes, these matches are likely to correspond to repetitive sequences such as transposable elements. Other matches form intermittent diagonal lines, and are likely to represent orthologous regions of the genome. When the angles of these diagonal lines change direction, they may represent “inversion” polymorphisms. This type of polymorphism occurs, when a large section of a chromosome flips-over (inverts), due to a breakage and erroneous repair.</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
You may need to load the R/4.0.2 module, then open R and install the “optparse” library using the install.packages command. Then exit R and run the script from the terminal.
</div>
</div>
<div class="callout-body-container callout-body">

</div>
</div>
<pre><code>Rscript ~/Workshop_materials/pafCoordsDotPlotly.R -i ColCroc_AriAges.paf -o ColCroc_AriAges.out -s -t -m 1500 -q 500000</code></pre>
</section>
<section id="rna-seq-gene-annotation" class="level2" data-link="6. More complex genomes">
<h2 data-link="6. More complex genomes" class="anchored" data-anchor-id="rna-seq-gene-annotation">8. RNA-seq gene annotation</h2>
<p>Earlier, we used some different annotation methods to model some genes found in a simple bacterial genome. Now we are going to have a go at using additional data to find genes in a more complicated organism. We will be using a very recent high quality genome for the diamondback moth.</p>
<section id="data-download" class="level3">
<h3 class="anchored" data-anchor-id="data-download">8.1. Data download</h3>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
To save time, you can skip downloading the data, and find it here: /home/taller-2019/Workshop_materials/
</div>
</div>
<div class="callout-body-container callout-body">

</div>
</div>
<p>First let’s download the genome data:</p>
<pre><code>wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/932/276/165/GCF_932276165.1_ilPluXylo3.1/GCF_932276165.1_ilPluXylo3.1_genomic.fna.gz</code></pre>
<p>and also some RNA seq data (It is pre-downloaded in /home/taller-2019/Workshop_materials/). The tissue that RNA is sampled from is very important. RNA composition varies very dramatically between tissues, developmental stages and environmental conditions. This sample comes from an adult moth, and therefore may not contain genes that are very specific to other stages, such as ovum, caterpillar &amp; pupae.</p>
<pre><code>fasterq-dump SRR15042536</code></pre>
</section>
<section id="mapping" class="level3">
<h3 class="anchored" data-anchor-id="mapping">8.2. Mapping</h3>
<p>Now in order to map the RNA-seq reads, we first need to index our genome.</p>
<p>First, decompress the genome:</p>
<pre><code>gunzip GCF_932276165.1_ilPluXylo3.1_genomic.fna.gz</code></pre>
<p>Then index like so:</p>
<pre><code>hisat2-build GCF_932276165.1_ilPluXylo3.1_genomic.fna GCF_932276165.1_ilPluXylo3.1_genomic_hisat2_index</code></pre>
<p>After a few minutes, we should have our genome indexed and ready for mapping. We can map our RNA-seq reads like this:</p>
<pre><code>hisat2 -S SRR15042536.sam -p 10 --rna-strandness=FR -x GCF_932276165.1_ilPluXylo3.1_genomic_hisat2_index -1 SRR15042536_1.fastq -2 SRR15042536_2.fastq --summary-file SRR15042536.summary</code></pre>
<p>Now we need to sort the resulting file and convert it to a compressed bam format (SAM is the human-readable version). We can also index the resulting bam, for visualisation purposes.</p>
<pre><code>samtools sort -o SRR15042536.sorted.bam SRR15042536.sam
samtools index SRR15042536.sorted.bam</code></pre>
</section>
<section id="transcript-assembly" class="level3">
<h3 class="anchored" data-anchor-id="transcript-assembly">8.3. Transcript assembly</h3>
<p>Now that we have our mapped reads, we can use the information to produce some gene models.</p>
<pre><code>stringtie -o SRR15042536.gff SRR15042536.sorted.bam</code></pre>
<p>You can now have a look at the annotation you’ve produced as well as the mapped read and see how the program is able to use RNA alignment evidence to build gene models.</p>
<p>IGV may not run on Kabre, but you can install it on your local machine (<a href="https://software.broadinstitute.org/software/igv/download">here</a>), or try the web app (<a href="https://igv.org/app/">here</a>):</p>
<pre><code>igv</code></pre>
<p>This should open a new GUI window. First load your genome, by clicking Genomes -&gt; Load Genome From File. Then you can load the sorted.bam file and the gff, by clicking File -&gt; Load from file.</p>
</section>
</section>
<section id="homology-gene-annotation" class="level2" data-link="6. More complex genomes">
<h2 data-link="6. More complex genomes" class="anchored" data-anchor-id="homology-gene-annotation">9. Homology gene annotation</h2>
<p>An alternative way to infer the presence of a gene is to use amino-acid sequences from closely related organisms. For many important protein sequences, purifying selection eliminates mutations that alter the amino-acid sequence, however coding sequences are still able to accumulate mutations in certain positions due to the fact that more multiple codon sequences are able to code for the same amino-acid.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Screenshot%202022-11-02%20at%2015.02.13.png" class="img-fluid figure-img"></p>
</figure>
</div>
<p>We will align some proteins for a very fast evolving gene and see how they compare to the RNA-seq data you’ve mapped. Here’ we’ll take a selection of the same protein from different organisms and see how their different levels of evolutionary divergence interferes with the alignment process.</p>
<p>Aligning proteins with miniprot is very simple, the minimal command is:</p>
<pre><code>miniprot -t8 --gff GCF_932276165.1_ilPluXylo3.1_genomic.fna Masc_multiSpecies_aa.fasta &gt; prot1.gff</code></pre>
<p>We can also tweak some of the parameters, and see how it affects the alignment:</p>
<pre><code>miniprot  -n 3 -s 3 -k 5 -l 3 -j 1 -t8 --gff GCF_932276165.1_ilPluXylo3.1_genomic.fna Masc_multiSpecies_aa.fasta &gt; prot2.gff</code></pre>
</section>
<section id="homology-gene-clustering" class="level2" data-link="6. More complex genomes">
<h2 data-link="6. More complex genomes" class="anchored" data-anchor-id="homology-gene-clustering">10. Homology gene clustering</h2>
<p>The determination of shared ancestry between genes in divergent species is a complex process. Reproductive isolation and other processes lead to varying levels of sequence divergence along different branches of an evolutionary tree. Genes may also become duplicated or rearranged, gain or lose specific functions. Interestingly, the evolutionary history of individual genes, may be radically different from a species tree, due to processes such as Incomplete lineage sorting, trans-species polymorphism (e.g.&nbsp;via balancing selection) and molecular convergence (see <a href="https://github.com/davidemms/OrthoFinder#orthogroups-orthologs--paralogs">here</a>). Thus, we may need to compare the trees of many genes to assess the true evolutionary tree of the species themselves.</p>
<p>A useful program <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1832-y">orthofinder</a>, provides many of these functions in an automated tool. Let’s download some protein data for five of the butterfly species in the table above (more species will take more time to run).</p>
<ol type="1">
<li><p>First make a new directory called orthofinder_1 and move into that folder.</p>
<pre><code>mkdir orthofinder_1
cd orthofinder_1</code></pre></li>
<li><p>Then, navigate to the FTP site for each species using the links above, then right click and copy the link to the “_protein.faa.gz” files. Use the wget command to copy the data into the new folder like so:</p>
<pre><code>wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/905/220/565/GCF_905220565.1_ilMelCinx1.1/GCF_905220565.1_ilMelCinx1.1_protein.faa.gz</code></pre></li>
<li><p>Decompress the protein files like this:</p>
<pre><code>GCF_905220565.1_ilMelCinx1.1_protein.faa.gz</code></pre></li>
<li><p>Once all the files are decompressed, you can run orthofinder like this:</p>
<pre><code>orthofinder -f ~/orthofinder_1</code></pre></li>
<li><p>This might take a little bit of time (there is a pre-computed output folder in the /home/taller-2019/Workshop_materials/ folder), but once complete, it is worth looking through the resulting output files (see <a href="https://davidemms.github.io/orthofinder_tutorials/exploring-orthofinders-results.html">here</a>). In particular you should now have a resolved species tree, and also trees for all genes in the inferred orthogroups. A simple way of viewing this data is to paste it into this <a href="http://etetoolkit.org/treeview/">website</a>.</p></li>
</ol>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>